state-network.md

```diff
@@ -125,61 +83,186 @@ A node should track their own radius value and provide this value in all Ping or
+
+```
+nibble     := {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, e, f}
+NibblePair := Byte  # 2 nibbles tightly packed into a single byte
+Nibbles    := Container(is_odd_length=bool, packed_nibbles=List(NibblePair, max_length=32))
+```
+
+`NibblePair` is packed so that nibble that goes first in the path uses high bits, while later nibble uses lower bits.
+
+`Nibbles.packed_nibbles` is a sequence of bytes with each byte containing two nibbles.  When encoding an odd length sequence of nibbles, `Nibbles.is_odd_length` boolean flag MUST be set to `True`, the high bits of the first byte MUST be left empty, first nibble MUST use low bits of the first byte, and remaning nibbles are tightly packed in remaining bytes.
+
+Examples:
+
+```
+[1, 2, a, b] -> Nibbles(is_odd_length=false, packed_nibbles=[0x12, 0xab])
+[1, 2, a, b, c] -> Nibbles(is_odd_length=true, packed_nibbles=[0x01, 0x2a, 0xbc])
+```
+
+##### Merkle Patricia Trie (MPT) Proofs
+
+Merkle Patricia Trie (MPT) proofs consist of a list of `WitnessNode` objects that correspond to individual trie nodes from the MPT. Each node can be one of the different node types from the MPT (e.g.blank, branch, extension, leaf).  When serialized, each `WitnessNode` is represented as an RLP serialized list of the component elements. The largest possible node type is the branch node which when serialized is a list of up to sixteen hashes in `Bytes32` (representing the hashes of each of the 16 nodes in that branch and level of the tree) plus the 4 elements of the node's value (balance, nonce, codehash, storageroot) represented as `Bytes32`.  When combined with the RLP prefixes, this yields a possible maximum length of 667 bytes.  We specify 1024 as the maximum length due to constraints in the SSZ spec for list lengths being a power of 2 (for easier merkleization.)
 
-#### Proofs
-Merkle Patricia Trie (MPT) proofs consist of a list of witness nodes that correspond to each trie node that consists of various data elements depending on the type of node (e.g.blank, branch, extension, leaf).  When serialized, each witness node is represented as an RLP serialized list of the component elements with the largest possible node type being the branch node which when serialized is a list of up to sixteen hashes in `Bytes32` (representing the hashes of each of the 16 nodes in that branch and level of the tree) plus the 4 elements of the node's value (balance, nonce, codehash, storageroot) represented as `Bytes32`.  When combined with the RLP prefixes, this yields a possible maximum length of 667 bytes.  We specify 1024 as the maximum length due to constraints in the SSZ spec for list lengths being a power of 2 (for easier merkleization.)
 ```
 WitnessNode            := ByteList(1024)
-MPTWitness             := List(witness: WitnessNode, max_length=32)
+Witness                := List(WitnessNode, max_length=1024)
```

Comment:
I think that this `max_length` is wrong.

I think the maximum should be 65 (assuming that there is a node on every level, which isn't realistic), while practically something way less should work (but 1024 is definitely too much).

Or am I missing something?